src/decoder/lattice-faster-decoder-combine-bucketqueue.cc

```diff
@@ -0,0 +1,1133 @@
+      return;
+    } else if (fst_->Type() == "vector") {
+      LatticeFasterDecoderCombineTpl<fst::VectorFst<fst::StdArc>, Token> *this_cast =
+          reinterpret_cast<LatticeFasterDecoderCombineTpl<fst::VectorFst<fst::StdArc>, Token>* >(this);
+      this_cast->AdvanceDecoding(decodable, max_num_frames);
+      return;
+    }
+  }
+
+
+  KALDI_ASSERT(!active_toks_.empty() && !decoding_finalized_ &&
+               "You must call InitDecoding() before AdvanceDecoding");
+  int32 num_frames_ready = decodable->NumFramesReady();
+  // num_frames_ready must be >= num_frames_decoded, or else
+  // the number of frames ready must have decreased (which doesn't
+  // make sense) or the decodable object changed between calls
+  // (which isn't allowed).
+  KALDI_ASSERT(num_frames_ready >= NumFramesDecoded());
+  int32 target_frames_decoded = num_frames_ready;
+  if (max_num_frames >= 0)
+    target_frames_decoded = std::min(target_frames_decoded,
+                                     NumFramesDecoded() + max_num_frames);
+  while (NumFramesDecoded() < target_frames_decoded) {
+    if (NumFramesDecoded() % config_.prune_interval == 0) {
+      PruneActiveTokens(config_.lattice_beam * config_.prune_scale);
+    }
+    ProcessForFrame(decodable);
+  }
+}
+
+// FinalizeDecoding() is a version of PruneActiveTokens that we call
+// (optionally) on the final frame.  Takes into account the final-prob of
+// tokens.  This function used to be called PruneActiveTokensFinal().
+template <typename FST, typename Token>
+void LatticeFasterDecoderCombineTpl<FST, Token>::FinalizeDecoding() {
+  ProcessNonemitting(NULL);
+  int32 final_frame_plus_one = NumFramesDecoded();
+  int32 num_toks_begin = num_toks_;
+  // PruneForwardLinksFinal() prunes final frame (with final-probs), and
+  // sets decoding_finalized_.
+  PruneForwardLinksFinal();
+  for (int32 f = final_frame_plus_one - 1; f >= 0; f--) {
+    bool b1, b2; // values not used.
+    BaseFloat dontcare = 0.0; // delta of zero means we must always update
+    PruneForwardLinks(f, &b1, &b2, dontcare);
+    PruneTokensForFrame(f + 1);
+  }
+  PruneTokensForFrame(0);
+  KALDI_VLOG(4) << "pruned tokens from " << num_toks_begin
+                << " to " << num_toks_;
+}
+
+
+template <typename FST, typename Token>
+void LatticeFasterDecoderCombineTpl<FST, Token>::ProcessForFrame(
+    DecodableInterface *decodable) {
+  KALDI_ASSERT(active_toks_.size() > 0);
+  int32 frame = active_toks_.size() - 1; // frame is the frame-index
+                                         // (zero-based) used to get likelihoods
+                                         // from the decodable object.
+  active_toks_.resize(active_toks_.size() + 1);
+
+  prev_toks_.swap(cur_toks_);
+  cur_toks_.clear();
+  if (prev_toks_.empty()) {
+    if (!warned_) {
+      KALDI_WARN << "Error, no surviving tokens on frame " << frame;
+      warned_ = true;
+    }
+  }
+
+  KALDI_ASSERT(best_token_in_next_frame_);
```

Comment:
You don't need best_token_in_next_frame_; get rid of all instances of it.  Instead you can work out the best cost on the current frame from the value `next_cutoff` had when you were doing ProcessForFrame() last time.
You can set up the code so that each time it processes a frame, it computes the cost_offset for the *next* frame, e.g., later on in this function (before updating adapative_beam), do
  ```
  BaseFloat best_cost_on_next_frame = next_cutoff - adaptive_beam;
  cost_offsets_.resize(frame + 2);
  cost_offsets_[frame + 1] = cost_offset;
```
Obviously you will have to initialize cost_offsets_ to size 1 with first element zero, in the function where you initialize everything.